### PR TITLE
Added support for single_border_width and single_margin for VerticalTile layout

### PR DESCRIPTION
**Pull request by matthewsia98**
Copy of previous PR because I had previously pushed to my fork's master branch.

Added single_border_width and single_margin keys for VerticalTile layout.

### CODE REVIEW

**Comment by elParaguayo** [comment]
I would still like a test added for this (in `test/layouts/test_verticaltile.py`. If you're not comfortable doing that then we can add it.

**Comment by matthewsia98** [comment]
> I would still like a test added for this (in `test/layouts/test_verticaltile.py`. If you're not comfortable doing that then we can add it.

I have added the tests for the new features.
I also went ahead and fixed the other tests. They were failing because they did not account for the default ```border_width = 1```.

File: test/layouts/test_verticaltile.py [hunk 1]
```diff
@@ -41,20 +41,31 @@ class VerticalTileConfig(Config):
...
 verticaltile_config = pytest.mark.parametrize("manager", [VerticalTileConfig], indirect=True)
+verticaltile_single_border_config = pytest.mark.parametrize("manager", [VerticalTileSingleBorderConfig], indirect=True)
+verticaltile_single_margin_config = pytest.mark.parametrize("manager", [VerticalTileSingleMarginConfig], indirect=True)
 
 
 @verticaltile_config
 def test_verticaltile_simple(manager):
     manager.test_window("one")
-    assert_dimensions(manager, 0, 0, 800, 600)
+    assert_dimensions(manager, 0, 0, 798, 598)
```
Comment (elParaguayo) [inline test/layouts/test_verticaltile.py:68]: Why has this test changed?

**Comment by elParaguayo** [comment]
> > I would still like a test added for this (in `test/layouts/test_verticaltile.py`. If you're not comfortable doing that then we can add it.
> 
> I have added the tests for the new features. I also went ahead and fixed the other tests. They were failing because they did not account for the default `border_width = 1`.

They weren't failing before...

**Comment by matthewsia98** [comment]
> > > I would still like a test added for this (in `test/layouts/test_verticaltile.py`. If you're not comfortable doing that then we can add it.
> > 
> > 
> > I have added the tests for the new features. I also went ahead and fixed the other tests. They were failing because they did not account for the default `border_width = 1`.
> 
> They weren't failing before...

Ah I know what's going on. 
Before adding the ```single_border_width``` option, it would effectively behave like ```single_border_width = 0```.

Here is the relevant code on master in verticaltile.py
```            
# width
if n > 1:
    width = screen_rect.width - self.border_width * 2
else:
    width = screen_rect.width
```

But in contrast, if we take a look at the MonadTall layout, we can see that if ```single_border_width``` is not specified, it will effectively behave like ```single_border_width = border_width```.

image

So I guess there needs to be a decision made on how VerticalTile will behave if ```single_border_width``` is not specified.

**Comment by elParaguayo** [comment]
There should be no change for users i.e. without setting the parameters that you've added, the behaviour should be identical to the current behaviour unless it's clearly wrong.

**Comment by matthewsia98** [comment]
Alright I have reverted the changes to the original tests and made the necessary fixes to keep old behavior.

I would like to just point out the inconsistency in behavior between ```VerticalTile``` and ```MonadTall``` again.

For ```MonadTall```, if we don't specify the ```single_border_width``` or ```single_margin``` keys, those keys will get their default values of ```None``` replaced with the values of ```border_width``` and ```margin``` respectively.

For ```VerticalTile```, if we don't specify the ```single_border_width``` or ```single_margin``` keys, those keys will keep their default value of ```None``` effectively setting them to 0.


As for which one is "correct" behavior, that's not for me to decide.

**Comment by elParaguayo** [comment]
Thanks. Inconsistency between different layouts is not ideal and should be something we look at. However, consistency in the behaviour of a layout when adding new features is important too.

Have restarted the tests so let's see how they get on.

**Comment by elParaguayo** [comment]
You need to run `black` on your changed files as there are a couple of formatting aspects causing our checks to fail.

File: test/layouts/test_verticaltile.py [hunk 2]
```diff
@@ -90,3 +101,23 @@ def test_verticaltile_window_focus_cycle(manager):
 
     # assert window focus cycle, according to order in layout
     assert_focus_path(manager, "float1", "float2", "one", "two", "three")
+
+
+@verticaltile_single_border_config
+def test_verticaltile_single_border(manager):
+    manager.test_window("one")
+    assert_dimensions(manager, 0, 0, 796, 596)
```
[the remark this answers is not in the document]
Comment (matthewsia98) [reply test/layouts/test_verticaltile.py:109]: Here's how it renders on my end.

Single Border
image

Normal Border
image

[the remark this answers is not in the document]
Comment (matthewsia98) [reply test/layouts/test_verticaltile.py:113]: Not sure what the code does exactly, but judgying by the presence of `outer_w` and `outer_h`, it does seem like borders are painted outside the actual window

image

[the remark this answers is not in the document]
Comment (m-col) [reply test/layouts/test_verticaltile.py:113]: Yes exactly right, the width and height are 'internal' dimensions that demarcate the contents of the window i.e. where the client renders. That would then be what `info()` reports in the tests too. It is a bit weird I agree, however these are the values we're sending in configure requests to the X server, meaning they are essentially X protocol abstractions. Of course that doesn't mean our `info()` command needs to align with that.

**Review by m-col — CHANGES_REQUESTED** [review]
> For `MonadTall`, if we don't specify the `single_border_width` or `single_margin` keys, those keys will get their default values of `None` replaced with the values of `border_width` and `margin` respectively.
> 
> For `VerticalTile`, if we don't specify the `single_border_width` or `single_margin` keys, those keys will keep their default value of `None` effectively setting them to 0.

As the latter are new additions, I think it would be preferable to implement them in line with the existing options of `MonadTall`. It's also consistent with some other options e.g. `margin_on_single` in `Columns`, which if left as its default value of `None` is replaced by the less specific `margin` option's value.

**Review by m-col — APPROVED** [review]
All LGTM thanks! I've enabled auto-merge so it will merge if CI passes.

**Pull request merged**